opt: use join filters to imply IS NOT NULL partial index predicates

mgartner opened the pull request:
#### opt: use join filters to imply IS NOT NULL partial index predicates

This commit updates the reject null normalization rules so that:

1. `RejectNullsUnderJoinRight` matches `SemiJoin` and `AntiJoin`.
2. A `Scan` with a `col IS NOT NULL` predicate requests null rejection
   of `col`.

In combination, these two changes allow partial indexes with
`IS NOT NULL` predicates to be used in more cases.

As an example, they can be used in JOINs where the ON condition
implicitly implies the predicate:

    CREATE TABLE a (a INT PRIMARY KEY)
    CREATE TABLE b (b INT, INDEX (b) WHERE b IS NOT NULL)

    SELECT * FROM a JOIN b ON a = b

As a another example, partial indexes with `IS NOT NULL` predicates can
be used to satisfy foreign key checks.

    CREATE TABLE parent (id INT PRIMARY KEY)

    CREATE TABLE child (
      id INT PRIMARY KEY,
      p_id INT,
      CONSTRAINT fk FOREIGN KEY (p_id) REFERENCES parent(id),
      INDEX (p_id) WHERE p_id IS NOT NULL
    )

    DELETE FROM p WHERE id = 1

The `DELETE` requires a foreign key check to ensure that no existing
rows in `child` have a matching `p_id`. Prior to this commit, the check
performed a full table scan rather than using the partial index because
its predicate was not implied. By pushing down the null-rejecting filter
derived from the check's `child.p_id = parent.id` filter, the partial
index can be used.

Fixes #57841

Release justification: This improves performance for joins and foreign
key checks on tables with partial indexes. Partial indexes are a new
feature introduced in 20.2.

Release note (performance improvement): Partial indexes with IS NOT NULL
predicates can be used in cases where JOIN filters implicitly imply the
predicate. This results in more efficient query plans for JOINs and
foreign checks.

mgartner committed:
opt: use join filters to imply IS NOT NULL partial index predicates

This commit updates the reject null normalization rules so that:

1. `RejectNullsUnderJoinRight` matches `SemiJoin` and `AntiJoin`.
2. A `Scan` with a `col IS NOT NULL` predicate requests null rejection
   of `col`.

In combination, these two changes allow partial indexes with
`IS NOT NULL` predicates to be used in more cases.

As an example, they can be used in JOINs where the ON condition
implicitly implies the predicate:

    CREATE TABLE a (a INT PRIMARY KEY)
    CREATE TABLE b (b INT, INDEX (b) WHERE b IS NOT NULL)

    SELECT * FROM a JOIN b ON a = b

As a another example, partial indexes with `IS NOT NULL` predicates can
be used to satisfy foreign key checks.

    CREATE TABLE parent (id INT PRIMARY KEY)

    CREATE TABLE child (
      id INT PRIMARY KEY,
      p_id INT,
      CONSTRAINT fk FOREIGN KEY (p_id) REFERENCES parent(id),
      INDEX (p_id) WHERE p_id IS NOT NULL
    )

    DELETE FROM p WHERE id = 1

The `DELETE` requires a foreign key check to ensure that no existing
rows in `child` have a matching `p_id`. Prior to this commit, the check
performed a full table scan rather than using the partial index because
its predicate was not implied. By pushing down the null-rejecting filter
derived from the check's `child.p_id = parent.id` filter, the partial
index can be used.

Fixes #57841

Release justification: This improves performance for joins and foreign
key checks on tables with partial indexes. Partial indexes are a new
feature introduced in 20.2.

Release note (performance improvement): Partial indexes with IS NOT NULL
predicates can be used in cases where JOIN filters implicitly imply the
predicate. This results in more efficient query plans for JOINs and
foreign checks.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58204)
<!-- Reviewable:end -->

mgartner commented on pkg/sql/opt/testutils/opttester/opt_tester.go:
```diff
@@ -661,14 +663,14 @@ func fillInLazyProps(e opt.Expr) {
 
 		// Derive columns that are candidates for null rejection.
-		norm.DeriveRejectNullCols(rel)
+		norm.DeriveRejectNullCols(md, rel)
```
I'm open to better ideas on how to pipe the metadata here, if you folks have any.

RaduBerinde reviewed (commented):
Thanks for working on this! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-:-MPBCO-DAgQdXAXVCndk:b7nhl9q)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/norm/reject_nulls_funcs.go, line 333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-MPBCal77WgeFwZL6YjU:-MPBAnP53_s1-3ituetR:beqf6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb157a1f280a54419ffbfe79d2f16af2a6dcc82a/pkg/sql/opt/norm/reject_nulls_funcs.go#L333)):*
> ```Go
> 
> 	var rejectNullCols opt.ColSet
> 	for _, pred := range md.TableMeta(scan.Table).PartialIndexPredicates {
> ```

We should be able to use `in.Memo().Metadata()` here instead of plumbing.
___
*[pkg/sql/opt/testutils/opttester/opt_tester.go, line 666 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-MPBAJgFcc2PL2zmLOec-r1-666:-MPBBhzD3wJg1rFNCdW_:b-67dcg9) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb157a1f280a54419ffbfe79d2f16af2a6dcc82a/pkg/sql/opt/testutils/opttester/opt_tester.go#L666)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm open to better ideas on how to pipe the metadata here, if you folks have any.
</blockquote></details>

I don't see a better way..
___
*[pkg/sql/opt/testutils/opttester/opt_tester.go, line 1568 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-MPBBs6L9UcjzazUmW0y:-MPBBs6L9UcjzazUmW0z:bkbjcua) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb157a1f280a54419ffbfe79d2f16af2a6dcc82a/pkg/sql/opt/testutils/opttester/opt_tester.go#L1568)):*
> ```Go
> }
> 
> func (ot *OptTester) optimizeExpr(o *xform.Optimizer) (opt.Expr, *opt.Metadata, error) {
> ```

[nit] This is a bit odd, since the caller is passing in the Optimizer (with which the Metadata is associated). I'd leave this unchanged and use `o.Factory().Metadata()` at the call site.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58204)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/norm/reject_nulls_funcs.go, line 333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-MPBCal77WgeFwZL6YjU:-MPBSRwG8q9dtzjfaAZC:b-t8zwew) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb157a1f280a54419ffbfe79d2f16af2a6dcc82a/pkg/sql/opt/norm/reject_nulls_funcs.go#L333)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We should be able to use `in.Memo().Metadata()` here instead of plumbing.
</blockquote></details>

Lol this is the "better way"! Now we don't need to make any opttester changes! :tada: 


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 6 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-:-MPCKdPC4owpLTVVvWSA:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @mgartner, and @rytaft)

___
*[pkg/sql/opt/exec/execbuilder/testdata/partial_index_nonmetamorphic, line 971 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-MPCK_99DFmmVG1wNBvz:-MPCK_99DFmmVG1wNBw-:bz0auhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dacf852e6107090285c8bce789d94fee32497f5/pkg/sql/opt/exec/execbuilder/testdata/partial_index_nonmetamorphic#L971)):*
> ```
> Del /Table/57/1/20/0
> 
> # ---------------------------------------------------------
> ```

I think this should go into a `partial_index` file since it doesn't involve traces.


<!-- Sent from Reviewable.io -->

mgartner committed:
opt: move execbuilder partial index EXPLAIN tests to metamorphic file

These tests do not show traces so they do not need to live in a
non-metamorphic test file.

Release note: None

mgartner committed:
opt: add partial index lookup join EXPLAIN test

This commit adds a test that verifies that the `(partial index)` label
is shown in `EXPLAIN` output when the query plan contains a lookup join
on a partial index.

Release note: None

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58204)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/exec/execbuilder/testdata/partial_index_nonmetamorphic, line 971 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-MPCK_99DFmmVG1wNBvz:-MPFtyHE2XwBoDwXJMWE:b-7s49rp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dacf852e6107090285c8bce789d94fee32497f5/pkg/sql/opt/exec/execbuilder/testdata/partial_index_nonmetamorphic#L971)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think this should go into a `partial_index` file since it doesn't involve traces.
</blockquote></details>

Done. Moved over the EXPLAIN tests too in a separate commit.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-:-MPGa6h7EWDG3gW41XV5:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @mgartner, @RaduBerinde, and @rytaft)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58204#-:-MPH3Xmx3CHZ5J5RnB6j:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @mgartner, @RaduBerinde, and @rytaft)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTR!

bors r=RaduBerinde
